### PR TITLE
register_route

### PR DESCRIPTION
**Pull request by roman454647**

### CODE REVIEW

File: routes/register.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    return errors;
+  }
+  const newUser = new User({
+    first_name: firstName,
+    last_name: lastName,
+    email,
+    password,
+  });
+
+  bcrypt.genSalt(10, (err, salt) => {
```
Comment (InnaKutsik) [inline routes/register.js]: pls move to separate function

File: routes/register.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    bcrypt.hash(newUser.password, salt, (err1, hash) => {
+      if (err1) {
+        console.log(err1);
+      }
+      newUser.password = hash;
+      newUser.save((error) => {
+        if (error) {
+          console.log(error);
```
Comment (InnaKutsik) [inline routes/register.js]: ?

File: routes/register.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    first_name: firstName,
+    last_name: lastName,
+    email,
+    password,
+  });
+
+  bcrypt.genSalt(10, (err, salt) => {
+    bcrypt.hash(newUser.password, salt, (err1, hash) => {
+      if (err1) {
+        console.log(err1);
```
Comment (InnaKutsik) [inline routes/register.js]: ?

File: routes/register.js (hunk 2 above)
Comment (roman454647) [reply routes/register.js]: created todo

File: routes/register.js (hunk 3 above)
Comment (roman454647) [reply routes/register.js]: created todo

File: routes/register.js (hunk 1 above)
Comment (roman454647) [reply routes/register.js]: moved

File: routes/register.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  }
+  const newUser = new User({
+    first_name: firstName,
+    last_name: lastName,
+    email,
+    password,
+  });
+  newUser.password = encrypt.encrypt(newUser);
+  newUser.save((error) => new Promise((resolve, reject) => {
+    setTimeout(() => {
```
Comment (InnaKutsik) [inline routes/register.js]: ?????? setTimeout

File: routes/encrypt.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+const bcrypt = require('bcryptjs');
+
+
+exports.encrypt = (newUser) => new Promise((resolve, reject) => {
+  setTimeout(() => {
```
Comment (InnaKutsik) [inline routes/encrypt.js]: setTimeout??

File: routes/register.js (hunk 4 above)
Comment (roman454647) [reply routes/register.js]: changed, save is already a promise in mongoose

File: routes/encrypt.js (hunk 5 above)
Comment (roman454647) [reply routes/encrypt.js]: deleted

File: routes/register.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    first_name: firstName,
+    last_name: lastName,
+    email,
+    password,
+  });
+  newUser.password = encrypt.encrypt(newUser);
+  newUser.save().then(
+    (error) => (new Error(error)),
+    () => (res.redirect('./login')),
+  );
```
Comment (InnaKutsik) [inline routes/register.js]: pls use chaining

it is more readable

```
newUser.save()
   .then(() => (res.redirect('/login'))
   .catch((error) => (new Error(error));

File: routes/encrypt.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const bcrypt = require('bcryptjs');
+
+
+exports.encrypt = (newUser) => new Promise((resolve, reject) => {
+  bcrypt.genSalt(10, (err, salt) => {
```
Comment (InnaKutsik) [inline routes/encrypt.js]: how about this? do we have promises from box?

File: routes/register.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+const User = require('../db/models/User');
+
+router.get('/', (req, res) => {
+  // TODO: render a register form
+  res.send('it is a register block');
+});
+
+
+router.post('/send', (req, res) => {
```
Comment (InnaKutsik) [inline routes/register.js]: pls create controller forlder and move there this callback

File: routes/register.js (hunk 6 above)
Comment (roman454647) [reply routes/register.js]: changed

File: routes/register.js (hunk 8 above)
Comment (roman454647) [reply routes/register.js]: created controller

File: routes/encrypt.js (hunk 7 above)
Comment (roman454647) [reply routes/encrypt.js]: yes

File: controllers/encrypt.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+const bcrypt = require('bcryptjs');
+
+exports.hashPassword = (password) => bcrypt.genSalt(10)
+  .then((salt) => bcrypt.hash(password, salt))
+  .then((hash) => hash)
+  .catch((err) => console.error(err.message));
```
Comment (InnaKutsik) [inline controllers/encrypt.js]: // TODO

File: controllers/register_controller.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+const mongoose = require('mongoose');
+const User = require('../db/models/User');
+
+const encrypt = require('./encrypt');
+
+exports.register_controller = (req, res) => {
+  const firstName = req.body.first_name;
+  const lastName = req.body.last_name;
+  const { email } = req.body;
+  const { password } = req.body;
```
Comment (InnaKutsik) [inline controllers/register_controller.js]: js destructurization
```
const { email, password, first_name: firstName } = req.bod;
```

File: controllers/register_controller.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        first_name: firstName,
+        last_name: lastName,
+        email,
+        password: result,
+      });
+      console.log(newUser);
+      newUser.save()
+        .then(() => (res.status(200).send('new user registered')))
+        .catch((error) => (res.send(error)));
+    });
```
Comment (InnaKutsik) [inline controllers/register_controller.js]:
```
return encrypt.hashPassword(password)
    .then((result) => {
      const newUser = new User({
        _id: new mongoose.Types.ObjectId().toHexString(),
        first_name: firstName,
        last_name: lastName,
        email,
        password: result,
      });
      return newUser.save();
    }).then(() => res.status(200).send('new user registered'))
    .catch((error) => res.send(error));
```

File: controllers/register_controller.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
+const mongoose = require('mongoose');
+const User = require('../db/models/User');
+
+const encrypt = require('./encrypt');
+
+exports.register_controller = (req, res) => {
```
Comment (InnaKutsik) [inline controllers/register_controller.js]: use camelCase

File: controllers/encrypt.js (hunk 9 above)
Comment (roman454647) [reply controllers/encrypt.js]: done

File: controllers/register_controller.js (hunk 10 above)
Comment (roman454647) [reply controllers/register_controller.js]: done

File: controllers/register_controller.js (hunk 11 above)
Comment (roman454647) [reply controllers/register_controller.js]: done

File: controllers/register_controller.js (hunk 12 above)
Comment (roman454647) [reply controllers/register_controller.js]: done

File: package.json [hunk 13]
```diff
@@ -24,14 +24,17 @@
...
+    "bcryptjs": "^2.4.3",
     "body-parser": "^1.19.0",
     "bootstrap": "^4.3.1",
     "dotenv": "^8.0.0",
     "express": "^4.17.1",
+    "express-validator": "^6.1.1",
     "mongodb": "^3.3.0",
     "mongoose": "^5.6.9",
+    "nodemon": "^1.19.1",
+    "passport": "^0.4.0"
```
Comment (anna-kuzii) [inline package.json]: add comma and we will not see red highlights

File: server.js [hunk 14]
```diff
@@ -1,8 +1,16 @@
...
 
 require('dotenv').config();
 
 const app = express();
+const dbConnect = require('./db/connect');
+
+dbConnect.connect();
+
+app.use(bodyParser.urlencoded({ extended: false }));
+// app.use(bodyParser.json());
```
Comment (anna-kuzii) [inline server.js]: do we need a comment here?

File: package.json (hunk 13 above)
Comment (roman454647) [reply package.json]: add

File: server.js (hunk 14 above)
Comment (roman454647) [reply server.js]: no

Comment (anna-kuzii) [reply server.js]: then rm it

Comment (roman454647) [reply server.js]: removed

**Pull request merged**